File: Euclid/Game/Plugins/PluginManager.cs
using System.Linq;
using System.Collections.Generic;
using System.IO;
using log4net;
using System.Reflection;

namespace Euclid.Game
{
    public class PluginManager : ILoadable
    {
        #region Fields

        public static readonly PluginManager Instance = new PluginManager();
        private List<PluginProxy> plugins;
        private string pluginDictionary = "plugins";

        private static readonly ILog log = LogManager.GetLogger(MethodBase.GetCurrentMethod().DeclaringType);

        #endregion

        #region Properties

        public IPlugin[] Plugins => plugins.Select(x => x.Plugin).ToArray();

        #endregion

        #region Constructors

        public void Load()
        {
            ReloadPlugins();
        }

        #endregion

        #region Public methods

        /// <summary>
        /// Reload plugin handler
        /// </summary>
        public void ReloadPlugins()
        {
            bool isReload = plugins != null;

            if (!Directory.Exists(pluginDictionary))
                Directory.CreateDirectory(pluginDictionary);

            if (plugins != null)
            {
                plugins.ForEach(x => x.Plugin.onDisable());
                plugins = null;
            }

            if (plugins == null)
            {
                plugins = new List<PluginProxy>();

                foreach (string file in Directory.GetFiles(pluginDictionary))
                {
                    if (Path.GetExtension(file) != ".dll")
                        continue;

                    log.Info($"Loading {Path.GetFileName(file)} as a plugin");
                    plugins.Add(new PluginProxy(file));
                }
            }

            plugins.ForEach(x => x.Plugin.onEnable());
        }

        #endregion
    }
}

File: euclid/src/main/java/euclid/game/plugins/PluginManager.java
package euclid.game.plugins;

import euclid.game.ILoadable;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

public class PluginManager implements ILoadable {

    private static final PluginManager instance = new PluginManager();
    private static final Logger log = LogManager.getLogger(PluginManager.class);
    private static final String PLUGIN_DIR = "plugins";

    private List<PluginProxy> plugins;

    private PluginManager() {
    }

    public static PluginManager getInstance() {
        return instance;
    }

    @Override
    public void load() {
        reloadPlugins();
    }

    public IPlugin[] getPlugins() {
        if (plugins == null) {
            return new IPlugin[0];
        }
        IPlugin[] result = new IPlugin[plugins.size()];
        for (int i = 0; i < plugins.size(); i++) {
            result[i] = plugins.get(i).getPlugin();
        }
        return result;
    }

    public void reloadPlugins() {
        boolean isReload = plugins != null;

        Path pluginPath = Paths.get(PLUGIN_DIR);
        if (!Files.exists(pluginPath)) {
            try {
                Files.createDirectories(pluginPath);
            } catch (IOException e) {
                log.error("Failed to create plugins directory", e);
                return;
            }
        }

        if (plugins != null) {
            for (PluginProxy proxy : plugins) {
                proxy.getPlugin().onDisable();
            }
            plugins = null;
        }

        plugins = new ArrayList<>();

        File[] files = new File(PLUGIN_DIR).listFiles();
        if (files != null) {
            for (File file : files) {
                if (!file.getName().endsWith(".jar")) {
                    continue;
                }
                log.info("Loading {} as a plugin", file.getName());
                try {
                    plugins.add(new PluginProxy(file.getAbsolutePath()));
                } catch (Exception e) {
                    log.error("Failed to load plugin {}", file.getName(), e);
                }
            }
        }

        for (PluginProxy proxy : plugins) {
            proxy.getPlugin().onEnable();
        }

        if (isReload) {
            log.info("Reloaded {} plugins", plugins.size());
        }
    }
}
